paddle/fluid/framework/prune.cc

```diff
@@ -269,124 +299,154 @@ void CloneWholeBlock(proto::ProgramDesc* input, proto::ProgramDesc* output,
+  // If not, the program need no pruning.
+  bool has_loss_grad_op = false;
+  std::queue<int> block_contains_loss;
+  std::queue<int> block_contains_loss_grad;
+  for (size_t i = 0; i < origin_clone.Size(); i++) {
+    auto block_ops = origin_clone.Block(i).AllOps();
+    for (auto op : block_ops) {
+      int op_role = boost::get<int>(
+          op->GetAttr(OpProtoAndCheckerMaker::OpRoleAttrName()));
+      if (op_role == (static_cast<int>(OpRole::kBackward) |
+                      static_cast<int>(OpRole::kLoss))) {
+        op->SetIsTarget(false);
+        has_loss_grad_op = true;
+        break;
+      }
     }
   }
-  PADDLE_ENFORCE_EQ(has_loss_op, true,
-                    "The Program need to be pruned its backward part"
-                    "should have loss operator.");
+  PADDLE_ENFORCE_EQ(
+      has_loss_grad_op, true,
+      platform::errors::InvalidArgument("The Program to be prune_backward"
+                                        "should have grad loss operator."));
 
-  // Step 2. Prune backward
+  // Step 2: Prune backward for each block.
   proto::ProgramDesc pruned_desc;
   pruned_desc.clear_blocks();
-  PruneBackwardImpl(origin_clone.Proto(), &pruned_desc, 0, -1);
+  std::map<int, int> pruned_progin_block_id_map;
+  for (size_t i = 0; i < origin_clone.Size(); i++) {
+    auto pruned = proto::BlockDesc();
+    auto origin = origin_clone.Proto()->mutable_blocks(i);
+
+    PruneBackwardImpl(origin, &pruned);
+    // If pruned block contains no operator, it means the block is a
+    // backward block and should be pruned.
+    // Else, add the block to pruned_desc and update its id & parent_id.
+    if (pruned.ops_size() > 0) {
+      auto* block_field = pruned_desc.mutable_blocks();
+      *block_field->Add() = pruned;
+
+      auto pruned_block_id = pruned_desc.blocks_size() - 1;
+      pruned_progin_block_id_map[pruned_block_id] = origin->idx();
+      auto* pruned_block = pruned_desc.mutable_blocks(pruned_block_id);
+      pruned_block->set_idx(pruned_block_id);
+
+      int parent_idx = -1;
+      if (origin->parent_idx() == -1) {
+        parent_idx = -1;
+      } else {
+        parent_idx =
+            FindMapByValue(pruned_progin_block_id_map, origin->parent_idx());
+        PADDLE_ENFORCE_NE(parent_idx, -1,
+                          "The parent block id should be found in "
+                          "pruned_progin_block_id_map");
+      }
+      pruned_block->set_parent_idx(parent_idx);
+    }
+  }
 
-  // Step 3. Contruct new framework::ProgramDesc
-  return std::unique_ptr<framework::ProgramDesc>(
-      new framework::ProgramDesc(pruned_desc));
-}
+  // Step 3: Update subblock attribute for conditional operator.
+  // This should be performed after all blocks pruned.
+  for (int i = 0; i < pruned_desc.blocks_size(); i++) {
+    auto* pruned = pruned_desc.mutable_blocks(i);
+    auto* ops = pruned->mutable_ops();
+    for (auto op_iter = ops->rbegin(); op_iter != ops->rend(); ++op_iter) {
```

Comment:
Why `rbegin`? It seems that `begin` is ok here.

Reply:
Done, thanks.